File: apps/ios/Sources/CoreData/RecordingDraftStore.swift
import Foundation
import CoreModel

/// A recording snapshot persisted across process death so a track survives an app kill.
/// Mirrors Android's `RecordingDraft`.
public struct RecordingDraft: Sendable, Equatable {
    public let points: [LatLng]
    public let elevations: [Double]
    public let elapsedSeconds: Int

    public init(points: [LatLng], elevations: [Double], elapsedSeconds: Int) {
        self.points = points
        self.elevations = elevations
        self.elapsedSeconds = elapsedSeconds
    }
}

/// Durable scratch store for the in-progress recording. The controller writes the accumulated
/// points (+ elevations + elapsed) as they grow and clears it on save/discard, so a relaunch
/// after a kill can resume the track. Mirrors Android's `RecordingDraftStore`.
public protocol RecordingDraftStore: Sendable {
    func load() async -> RecordingDraft?
    func save(points: [LatLng], elevations: [Double], elapsedSeconds: Int) async
    func clear() async
}

/// No-op store (tests / previews) — nothing survives a kill.
public struct NoopRecordingDraftStore: RecordingDraftStore {
    public init() {}
    public func load() async -> RecordingDraft? { nil }
    public func save(points: [LatLng], elevations: [Double], elapsedSeconds: Int) async {}
    public func clear() async {}
}

/// `UserDefaults`-backed store. Points/elevations are encoded as compact `;`-delimited strings
/// (same shape as Android's DataStore draft) so a recording survives process death.
public actor UserDefaultsRecordingDraftStore: RecordingDraftStore {
    private let defaults: UserDefaults
    private enum Key {
        static let points = "recording.draft.points"
        static let elevations = "recording.draft.elevations"
        static let elapsed = "recording.draft.elapsed"
    }

    public init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    public func load() async -> RecordingDraft? {
        guard let encoded = defaults.string(forKey: Key.points), !encoded.isEmpty else { return nil }
        let points = encoded.split(separator: ";").compactMap { pair -> LatLng? in
            let parts = pair.split(separator: ",")
            guard parts.count == 2, let lat = Double(parts[0]), let lng = Double(parts[1]) else { return nil }
            return LatLng(lat: lat, lng: lng)
        }
        guard !points.isEmpty else { return nil }
        let elevations = (defaults.string(forKey: Key.elevations) ?? "")
            .split(separator: ";", omittingEmptySubsequences: false)
            .compactMap { Double($0) }
        return RecordingDraft(points: points, elevations: elevations, elapsedSeconds: defaults.integer(forKey: Key.elapsed))
    }

    public func save(points: [LatLng], elevations: [Double], elapsedSeconds: Int) async {
        defaults.set(points.map { "\($0.lat),\($0.lng)" }.joined(separator: ";"), forKey: Key.points)
        defaults.set(elevations.map { String($0) }.joined(separator: ";"), forKey: Key.elevations)
        defaults.set(elapsedSeconds, forKey: Key.elapsed)
    }

    public func clear() async {
        defaults.removeObject(forKey: Key.points)
        defaults.removeObject(forKey: Key.elevations)
        defaults.removeObject(forKey: Key.elapsed)
    }
}

File: apps/android/core/tracking/src/main/kotlin/com/sigmundgranaas/turbo/expressive/core/tracking/RecordingDraftStore.kt
package com.sigmundgranaas.turbo.expressive.core.data

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.sigmundgranaas.turbo.expressive.domain.LatLng
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.first
import javax.inject.Inject
import javax.inject.Singleton

/**
 * A recording snapshot persisted across process death so a track survives a kill.
 * [pausedFromIndex] ≥ 0 means the session was paused: points before it are the track, points
 * from it on are the paused buffer (US-4) — so a kill-while-paused doesn't lose that walk.
 */
data class RecordingDraft(
    val points: List<LatLng>,
    val elevations: List<Double?>,
    val elapsedSec: Int,
    val pausedFromIndex: Int = -1,
)

/**
 * Durable scratch store for the in-progress recording. The controller writes the
 * accumulated points (+ elevations + elapsed time) as they grow and clears it on
 * save/discard, so a foreground-service restart (or process death) can resume the
 * track — elevation profile included.
 */
interface RecordingDraftStore {
    suspend fun load(): RecordingDraft?
    suspend fun save(points: List<LatLng>, elevations: List<Double?>, elapsedSec: Int, pausedFromIndex: Int = -1)
    suspend fun clear()
}

private val Context.recordingDraftStore: DataStore<Preferences> by preferencesDataStore(name = "recording_draft")

@Singleton
class DataStoreRecordingDraftStore @Inject constructor(
    @param:ApplicationContext private val context: Context,
) : RecordingDraftStore {

    private object Keys {
        val POINTS = stringPreferencesKey("draft_points")
        val ELEVATIONS = stringPreferencesKey("draft_elevations")
        val ELAPSED = intPreferencesKey("draft_elapsed")
        val PAUSED_FROM = intPreferencesKey("draft_paused_from")
    }

    override suspend fun load(): RecordingDraft? {
        val prefs = context.recordingDraftStore.data.first()
        val encoded = prefs[Keys.POINTS]?.takeIf { it.isNotBlank() } ?: return null
        val points = encoded.split(";").mapNotNull { pair ->
            val parts = pair.split(",")
            val lat = parts.getOrNull(0)?.toDoubleOrNull()
            val lng = parts.getOrNull(1)?.toDoubleOrNull()
            if (lat != null && lng != null) LatLng(lat, lng) else null
        }
        if (points.isEmpty()) return null
        // Elevations are parallel to points; an empty token means "no altitude".
        val elevations = prefs[Keys.ELEVATIONS]
            ?.split(";")
            ?.map { it.toDoubleOrNull() }
            ?.takeIf { it.size == points.size }
            ?: List(points.size) { null }
        return RecordingDraft(points, elevations, prefs[Keys.ELAPSED] ?: 0, prefs[Keys.PAUSED_FROM] ?: -1)
    }

    override suspend fun save(points: List<LatLng>, elevations: List<Double?>, elapsedSec: Int, pausedFromIndex: Int) {
        context.recordingDraftStore.edit { prefs ->
            prefs[Keys.POINTS] = points.joinToString(";") { "${it.lat},${it.lng}" }
            prefs[Keys.ELEVATIONS] = elevations.joinToString(";") { it?.toString().orEmpty() }
            prefs[Keys.ELAPSED] = elapsedSec
            prefs[Keys.PAUSED_FROM] = pausedFromIndex
        }
    }

    override suspend fun clear() {
        context.recordingDraftStore.edit { it.clear() }
    }
}
